src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2298,4 +2298,496 @@ - (void)testSubscriptionPool
+            for (NSNumber * endpoint in testEndpoints) {
+                XCTAssertNotNil(initialClusterIndex);
+                for (NSNumber * cluster in [initialClusterIndex objectForKey:endpoint]) {
+                    if ([endpoint isEqualToNumber:toBeDeletedEndpoint]) {
+                        XCTAssertNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                    } else {
+                        XCTAssertNotNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                    }
+                }
+            }
+        });
+        [reportEndExpectation fulfill];
+    };
+
+    [device unitTestInjectAttributeReport:attributeReport fromSubscription:YES];
+
+    [self waitForExpectations:@[ attributeDataReceivedExpectation, reportEndExpectation ] timeout:60];
+
+    [controller.controllerDataStore clearAllStoredClusterData];
+    NSDictionary * storedClusterDataAfterClear = [controller.controllerDataStore getStoredClusterDataForNodeID:deviceID];
+    XCTAssertEqual(storedClusterDataAfterClear.count, 0);
+
+    [controller removeDevice:device];
+    // Reset our commissionee.
+    __auto_type * baseDevice = [MTRBaseDevice deviceWithNodeID:deviceID controller:controller];
+    ResetCommissionee(baseDevice, queue, self, kTimeoutInSeconds);
+
+    [controller shutdown];
+    XCTAssertFalse([controller isRunning]);
+}
+
+- (void)test012_testDataStorageUpdatesWhenRemovingClusters
+{
+    NSNumber * deviceID = @(17);
+    __auto_type * device = [self getMTRDevice:deviceID];
+    __auto_type queue = dispatch_get_main_queue();
+    __auto_type * delegate = [[MTRDeviceTestDelegate alloc] init];
+    __auto_type * controller = device.deviceController;
+
+    XCTestExpectation * subscriptionExpectation = [self expectationWithDescription:@"Subscription has been set up"];
+
+    __block unsigned attributeReportsReceived = 0;
+    __block NSNumber * dataVersionForServerList;
+    __block NSNumber * testEndpoint = @1;
+
+    // This test will do the following -
+    // 1. Get the data version and attribute value of the server list for endpoint 1 to inject a fake report. The attribute report will delete cluster ID - MTRClusterIDTypeIdentifyID.
+    //    That should cause the cluster to be removed from cluster index for endpoint 1 and the cluster data for the removed cluster should be cleared from data storage.
+    // 2. The data store is populated with MTRClusterIDTypeIdentifyID in the cluster index and cluster data for endpoint 1 initially.
+    // 3. After the fake attribute report is injected with deleted cluster ID - MTRClusterIDTypeIdentifyID, make sure the data store is still populated with cluster index and
+    //    cluster data for all other clusters at endpoint 1 but not the deleted cluster.
+    __block NSMutableArray * testClusterDataValue;
```

Comment:
All the same comments as for the 011 test apply: types, unnecessary block vars, name of test function, etc.